components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt

```diff
@@ -97,27 +101,22 @@ class ToolbarPresenter(
 
         extensions.forEach { extension ->
             extension.browserAction?.let { extensionAction ->
-                val toolbarAction = convertToToolbarAction(extensionAction)
-                // We need to verify, if we this toolbarAction already exits on the toolbar
-                // if so, we don't need to re-add it, just replace it with the new one.
-                // https://github.com/mozilla-mobile/android-components/issues/4651
-                toolbar.addBrowserAction(toolbarAction)
+                val existingBrowserAction = webExtensionToolbarMap[extension.id]
+
+                if (existingBrowserAction != null) {
+                    existingBrowserAction.browserAction = extensionAction
+                } else {
+                    val toolbarAction = WebExtensionToolbarAction(
+                        browserAction = extensionAction,
+                        listener = extensionAction.onClick)
+                    toolbar.addBrowserAction(toolbarAction)
+                    webExtensionToolbarMap[extension.id] = toolbarAction
+                }
                 toolbar.invalidateActions()
```

Comment:
Initially, I thought we're gonna look for a way to only call invalidate on the single action that's affected. This will invalidate the menu/browserActions/pageActions/navigationActions every time. For uBlock this will definitely be called a lot, but I am not sure that's actually a performance problem. So, OK to land as is if it's tricky to implement and we can revisit later.

Reply:
@pocmo I remembered that we talked about this. But I don't recall if you have any concerns related to add functionality for updating  a specific `Toolbar.Action `. Should we worry about calling `toolbar.invalidateActions()` when just one item has changed?

Reply:
I think that is fine. On invalidate the toolbar will:
- call the visible lambda of all actions to check whether they still want to be in the toolbar
- (only if needed it adds or removes views for new/hidden/visible actions)
- it will call bind() on all actions to give them the opportunity to update their view (which gets passed in)

This is cheap: The visible lambda doesn't do much and the bind() call for all default actions doesn't do anything at all because by default actions just don't change.

Reply:
Thanks!
cc: @psymoon

Reply:
Applications may do expensive things in the various lambdas though which are called during `bind` e.g. see also https://github.com/mozilla-mobile/android-components/issues/4915 which is needed because a first version queried the bookmark state from storage in `TwoStateButton#isInPrimaryState`. Of course, we could document that this is a bad idea and hope apps don't do that...